### ISSUE TITLE
how should we understand the integrate flag of models?

### ISSUE THREAD

**Issue by DougBurke** [sherpa/sherpa#1801]
This is related to my attempts to improve our regrid code - e.g.

- #1482 
- #1799 

and other issues and PRs, but is a precursor to them (i.e. I strongly believe we need to answer this question before I can work on them).

Technically this is independent of model dimensionality, but I'm going to focus on the 1D case (because we don't really have tests of 2D integrated datasets).

For our models in `sherpa.models.basic`, the model evaluation is done in C++ scope, and in general has two routines for each model - a "point" model and an "integrated" model (there are some which use numerical integration to get the "integrated" model). These models are sent `(pars, xlo, xhi=None, integrate=True)` and they use the "point" model and send in the `xlo` values **unless** `xhi` is given and `integrate` is `True`, in which case they use the integrated version and send in `xlo` and `xhi`. This means that you can send a model `xlo` and `xhi` values but that the `xhi` values will be ignored unless `integrate` is `True` (which is the default value).

So, this appears to be one way of defining model behavior for 1D models

- you can be sent (pars, x, xhi=None) or (pars, x, xhi)
- you only ever use x **unless** xhi is not None AND self.integrate is True

although note that this logic is hidden in the C++ code. The Python layer will allow models to be called with `xhi` not set to None even when `self.integrate=False`.

However, you may want to do something a bit different. For example, I could imagine we'd want to use `(x + xhi) / 2` as the independent axis for a model when sent in an integrated grid (that is, xhi is not None), but the model has `integrate` set to False. Or, I may essentially ignore the `integrate` setting and have the `calc` method be something like

```python
    def calc(self, p, x, xhi=None, **kwargs):
        if xhi is None:
            xmid = x
        else
            xmid = (x + xhi) / 2
        return _do_something_with_xmid
```

This could be for a model where I would ideally want the `integrate` flag to always be `False`.

When we get to XSPEC we get into some fun. We always [*] send them `(x, xhi)` bins - e.g. send in an integrated grid - and it's up to the model itself what to do. The XSPEC additive models will integrate across the bin but the multiplicative models are free to use any energy within the bin (left value, mean, or perhaps just look to see if an edge happens within the bin). I plan - in #1179 or elsewhere - to make the XSPEC multiplicative models have `integrate=False`.

So what I'd like us to think about is that the `self.integrate` flag indicates whether we "integrate across" a bin. This only makes sense for integrated axes (where you give `x` and `xhi`). If `integrate=False` you can be sent `xhi` not equal to None. It is not a huge difference, but one that I want us to come to some understanding on - and because it'll then lead on to reviewing what I wrote in
- #1482

---

[*] actually, we can send in just an `x` array but it is treated as a set of consecutive "integrated" bins and this is only really of interest for people who want to call the model directly, and there's a warning that gets triggered when they do so

**Comment by hamogu** [comment]
I agree with the meaning of "integrate". "Integrate=True" makes sense if and only if xlo and xhi are given.
There should only be two cases:
- integrate=True and xlo and xhi given
- integrate=False, and only x given (xhi not given or None)

Now, we currently allow other combinations, but I think [*] that's more of a historical accident than a sound design, because we often call `calc(*args, **kwargs)`. In #1796 I suggested (and @DougBurke had suggested something like that earlier) a more structured way to pass in data. If we do, the integrate flag on the model itself is no longer necessary; the model will know if it's getting a integrated or a point dataset.

[*] At least that's my impression from reading code and comments; I have not seen the original design documents.

**Comment by hamogu** [comment]
Remark: 

> although note that this logic is hidden in the C++ code. 

It would be useful to lift a few (at least one per dimension) of the models into the Python scope. I refuse to believe that evaluating a constant in Python creates a speed bottleneck; on the other hand, having at least one real model in Python would provide an example to build on when reading the code to learn how it works.

**Comment by DougBurke** [comment]
@hamogu - your suggestion won't work. XSPEC multiplicative models take (xlo, xhi) values but do not integrate within a bin. The Sherpa Scale1D model is similar (see https://github.com/sherpa/sherpa/issues/713#issuecomment-1591737993). So I think there is value in having the `integrate` flag. I have also done a lot of radial fitting where we have a "histogram" dataset but have decided not to fit with integrated models to increase speed (and also because in this particular case integration across the bin requires knowledge of areas and so isn't as simple as the PHA case where you can just do $\int f(e) \rm{de}$ 

I also want to use it to understand how to do regridding - since there's a difference between flux conservation vs interpolation.

**Comment by hamogu** [comment]
I would be up to the model to decide what to do with that information. Code could still be
```Python
if 'x' in inputdata:
    return calc_absortion(x)
if 'xhi' in inputdata and 'xlo' in inputdata:
    return calc_absorption(xlo)
    # or use midpoint
    return calc_absorption((xlo + xhi) / 2)
    # or subsample
    absorp = np.zeros_like(xlo)
    for i in range(len(xlo)):
        absorp[i] = np.mean(calc_absorb(np.linspace(xlow, xhi, 10)))
   # or do full numerical integration
```

I agree that 2d models are more difficult, but then it might be clearer to represent your data as `Data2D` and not `Data2DInt` in your use case. Yes, the data is technically integrated, but in practice you want to treat it like point data, so the models should get it as point data. 
I guess, in the current paradigm, you would pass in `Data2DInt`, but then set the `integrate` flare got the models to `False`? That's certainly possible, but if we are rethinking this, it strikes me as a bit of a hack
